File: app/src/main/java/com/example/admin/weatherchallenge/view/weather/WeatherPresenter.java
package com.example.admin.weatherchallenge.view.weather;

import android.util.Log;

import com.example.admin.weatherchallenge.manager.DataRepository;
import com.example.admin.weatherchallenge.model.entity.WeatherResponse;
import com.example.admin.weatherchallenge.model.remote.WeatherCallback;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class WeatherPresenter implements WeatherContract.Presenter {
    private static String TAG = WeatherPresenter.class.getSimpleName() + "_TAG";
    private WeatherContract.View view;
    private DataRepository dataRepository;

    public WeatherPresenter(DataRepository dataRepository) {
        this.dataRepository = dataRepository;
    }

    @Override
    public void getWeather(@NotNull String city) {
        Log.d(TAG, "getWeather: " + city);
        dataRepository.getWeather(city, new WeatherCallback() {

            @Override
            public void onRemoteResponse(@NotNull List<WeatherResponse> weatherResponseList) {
                Log.d(TAG, "onRemoteResponse: " + weatherResponseList.size());
                view.onGetWeather(weatherResponseList);
            }

            @Override
            public void onRemoteFailure(@NotNull String error) {
                view.showError(error);
            }
        });
    }

    @Override
    public void attachView(@NotNull WeatherContract.View view) {
        this.view = view;
    }

    @Override
    public void detachView() {
        this.view = null;
    }
}

File: app/src/main/java/com/example/admin/weatherchallenge/view/weather/WeatherPresenter.kt
package com.example.admin.weatherchallenge.view.weather

import android.util.Log

import com.example.admin.weatherchallenge.manager.DataRepository
import com.example.admin.weatherchallenge.model.entity.WeatherResponse
import com.example.admin.weatherchallenge.model.remote.WeatherCallback

class WeatherPresenter(private val dataRepository: DataRepository) : WeatherContract.Presenter {
    private var view: WeatherContract.View? = null

    override fun getWeather(city: String) {
        Log.d(TAG, "getWeather: $city")
        dataRepository.getWeather(city, object : WeatherCallback {

            override fun onRemoteResponse(weatherResponseList: List<WeatherResponse>) {
                Log.d(TAG, "onRemoteResponse: " + weatherResponseList.size)
                view!!.onGetWeather(weatherResponseList)
            }

            override fun onRemoteFailure(error: String) {
                view!!.showError(error)
            }
        })
    }

    override fun attachView(view: WeatherContract.View) {
        this.view = view
    }

    override fun detachView() {
        this.view = null
    }

    companion object {
        private val TAG = WeatherPresenter::class.java.simpleName + "_TAG"
    }
}
